lib/NativeCall.pm6

```diff
@@ -588,6 +588,21 @@ multi sub postcircumfix:<[ ]>(CArray:D \array, $pos) is export(:DEFAULT, :types)
 multi sub postcircumfix:<[ ]>(CArray:D \array, *@pos) is export(:DEFAULT, :types) {
     @pos.map: { array.AT-POS($_) };
 }
+multi sub postcircumfix:<[ ]>(CArray:D \array, Callable:D $block) is export(:DEFAULT, :types) {
+    nqp::stmts(
+      (my $*INDEX = 'Effective index'),
+      array[$block.pos(array)]
+    )
+}
+multi sub postcircumfix:<[ ]>(CArray:D \array) is export(:DEFAULT, :types) {
+    array[^array.elems]
```

Comment:
In regular lists, this one has [explict "just return `self`" meaning](https://github.com/rakudo/rakudo/blob/72c8cf68c/src/core/Any.pm6#L303-L310), which, for example, avoids reification.

```
<Zoffix_> m: my @a := <a b c>; dd @a =:= @a[]
<camelia> rakudo-moar 0ec4f5f28: OUTPUT: «Bool::True␤»
<Zoffix_> m: my @a := <a b c>; dd @a =:= @a[*]
<camelia> rakudo-moar 0ec4f5f28: OUTPUT: «Bool::False␤»
```

Does it make sense for `CArray` to simply return the `array` we get as-is without any indexing into it?

Reply:
Could it be that the slicing is necessary to create true HLL Perl 6 objects?

Reply:
Looking at candidates, seems [the candidate that calls `.ZEN-POS` is already called](https://github.com/rakudo/rakudo/blob/0ec4f5f283905f4e4ca79cac6a92ed12ec74310e/src/core/array_slice.pm6#L478-L480), but you don't right the right thing when you, say, interpolate it in a string:

```
<Zoffix_> m: use NativeCall; my @z  := CArray[int32].new( 1, 2, 3, 4, 5 ); say "foo @z[] bar"
<camelia> rakudo-moar 0ec4f5f28: OUTPUT: «foo  bar␤»
```

Makes me think we should be adding methods to [CArray class](https://github.com/rakudo/rakudo/blob/0ec4f5f283905f4e4ca79cac6a92ed12ec74310e/lib/NativeCall/Types.pm6#L92) rather than adding more postcircumfix operators, since even with the postcircumfix defined, calling `.ZEN-POS` directly would still give the wrong thing.